### PR TITLE
Refactor Group and Profile selection into core server

### PR DESCRIPTION
**Pull request by dghubble**
* Add gRPC SelectGroup and SelectProfile service endpoints
* HTTP and gRPC servers should both allow labels to be resolved to the matching Group or Profile. Move logic to bootcfg core server
* Add gRPC wrapper selectServer to avoid cluttering core server with response types which are currently only useful for gRPC

### CODE REVIEW

**Comment by dghubble** [comment]
@sym3tri

File: bootcfg/api/handlers.go [hunk 1]
```diff
@@ -25,3 +30,37 @@ func logRequests(next http.Handler) http.Handler {
 	}
 	return http.HandlerFunc(fn)
 }
+
+// selectGroup selects the Group whose selectors match the query parameters,
+// adds the Group to the ctx, and calls the next handler. The next handler
+// should handle a missing Group.
```
Comment (sym3tri) [inline bootcfg/api/handlers.go]: nit: having comments here is great for GoDoc. It's also nice to have them in the proto file so the entire API is human parsable with 1 file. Maintaining 2 sets of identical comments is not good, so I propose putting them in just the proto file b/c the actual server implementation methods are unlikely to be dependencies of anything.

Comment (sym3tri) [inline bootcfg/api/handlers.go]: disregard. i thought this was the rpc handler.

Comment (dghubble) [inline bootcfg/api/handlers.go]: hm, selectGroup is a HTTP handler that deals with query parameters, contexts, and http.Requests. The proto service SelectGroup is the underlying logic which is shared.

Maybe within the server package I'll leave comments in the proto, but in `api` (probably should be called `www` or `web`) they'd make sense?

File: bootcfg/server/serverpb/rpc.proto [hunk 2]
```diff
@@ -19,6 +19,27 @@ service Profiles {
   rpc ProfileList(ProfileListRequest) returns (ProfileListResponse) {};
 }
 
+service Select {
+  rpc SelectGroup(SelectGroupRequest) returns (SelectGroupResponse) {};
```
Comment (sym3tri) [inline bootcfg/server/serverpb/rpc.proto]: What does "Selection" actually do?

Comment (dghubble) [inline bootcfg/server/serverpb/rpc.proto]: There should be several methods like these. SelectGroup, SelectProfile, SelectIgnition, SelectCloudConfig. Basically machine labels (usually from query parameters) are used to select the first matching Group, Profile, Ignition Config, etc.

File: bootcfg/rpc/select.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+func newSelectServer(s server.Server) pb.SelectServer {
+	return &selectServer{
+		srv: s,
+	}
+}
+
+func (s *selectServer) SelectGroup(ctx context.Context, req *pb.SelectGroupRequest) (*pb.SelectGroupResponse, error) {
+	group, err := s.srv.SelectGroup(ctx, req)
+	return &pb.SelectGroupResponse{Group: group}, err
```
Comment (sym3tri) [inline bootcfg/rpc/select.go]: Maybe in a separate PR...
it's best to use standard gRPC error codes with your returned errors.

https://godoc.org/google.golang.org/grpc#Errorf

File: bootcfg/api/handlers.go (hunk 1 above)
Comment (sym3tri) [inline bootcfg/api/handlers.go]: i think it's fine as-is. sorry my first comment was a mistake.

File: bootcfg/server/serverpb/rpc.proto (hunk 2 above)
Comment (sym3tri) [inline bootcfg/server/serverpb/rpc.proto]: WFM. Maybe brief comments would be good here.

**Comment by sym3tri** [comment]
otherwise LGTM

File: bootcfg/rpc/select.go (hunk 3 above)
Comment (dghubble) [inline bootcfg/rpc/select.go]: Ok, I'll do that in a followup soon

**Comment by dghubble** [comment]
Thanks

**Pull request merged**